File: ios/Shared/Sources/Networking/LmsWebServiceError.swift
import Foundation

public enum LmsWebServiceError: LocalizedError {
    case invalidSiteURL
    case invalidResponse
    case invalidSubmissionFiles
    case http(statusCode: Int, response: String)
    case moodle(message: String, debugInfo: String?)
    case decoding(Error)

    public var errorDescription: String? {
        switch self {
        case .invalidSiteURL:
            return "Moodle の API URL を組み立てられませんでした。"
        case .invalidResponse:
            return "Moodle から不正なレスポンスを受け取りました。"
        case .invalidSubmissionFiles:
            return "提出するファイルを選択してください。"
        case .http(let statusCode, let response):
            if response.isEmpty {
                return "Moodle が HTTP \(statusCode) を返しました。"
            }
            return "Moodle が HTTP \(statusCode) を返しました。 \(response)"
        case .moodle(let message, let debugInfo):
            guard let debugInfo, debugInfo.isEmpty == false else {
                return message
            }
            return "\(message) (\(debugInfo))"
        case .decoding:
            return "Moodle のレスポンスを解析できませんでした。"
        }
    }
}

File: android/networking/src/main/kotlin/dev/example/moodlenative/networking/LmsWebServiceError.kt
package dev.example.moodlenative.networking

sealed class LmsWebServiceError(
    override val message: String,
    override val cause: Throwable? = null,
) : Exception(message, cause) {
    data object InvalidSiteURL : LmsWebServiceError("Moodle の API URL を組み立てられませんでした。")

    data object InvalidResponse : LmsWebServiceError("Moodle から不正なレスポンスを受け取りました。")

    data object InvalidSubmissionFiles : LmsWebServiceError("提出ファイルが選択されていません。")

    data class Http(
        val statusCode: Int,
        val response: String,
    ) : LmsWebServiceError(
        if (response.isEmpty()) {
            "Moodle が HTTP $statusCode を返しました。"
        } else {
            "Moodle が HTTP $statusCode を返しました。 $response"
        },
    )

    data class Moodle(
        val moodleMessage: String,
        val debugInfo: String?,
    ) : LmsWebServiceError(
        if (debugInfo.isNullOrEmpty()) {
            moodleMessage
        } else {
            "$moodleMessage ($debugInfo)"
        },
    )

    data class Decoding(
        val decodingCause: Throwable,
    ) : LmsWebServiceError("Moodle のレスポンスを解析できませんでした。", decodingCause)
}
